tests/50federation/30room-join.pl

```diff
@@ -327,6 +326,227 @@ sub assert_is_valid_pdu {
+            assert_json_object( $auth_event->[1] );  # hashes
+         }
+
+         assert_json_nonempty_list( $protoevent->{prev_events} );
+
+         assert_json_number( $protoevent->{depth} );
+
+         $protoevent->{room_id} eq $room_id or
+            die "Expected 'room_id' to be $room_id";
+         $protoevent->{sender} eq $user_id or
+            die "Expected 'sender' to be $user_id";
+         $protoevent->{state_key} eq $user_id or
+            die "Expected 'state_key' to be $user_id";
+         $protoevent->{type} eq "m.room.member" or
+            die "Expected 'type' to be 'm.room.member'";
+
+         assert_json_keys( my $content = $protoevent->{content}, qw( membership ) );
+         $content->{membership} eq "join" or
+            die "Expected 'membership' to be 'join'";
+
+         my %event = (
+            ( map { $_ => $protoevent->{$_} } qw(
+               auth_events content depth prev_events room_id sender
+               state_key type ) ),
+
+            event_id         => $datastore->next_event_id,
+            origin           => $local_server_name,
+            origin_server_ts => $inbound_server->time_ms,
+         );
+
+         $datastore->sign_event( \%event );
+
+         $outbound_client->do_request_json(
+            method   => "PUT",
+            hostname => $first_home_server,
+            uri      => "/v2/send_join/$room_id/$event{event_id}",
+
+            content => \%event,
+         )
+      })->then( sub {
+         my ( $response ) = @_;
+
+         assert_json_keys( $response, qw( auth_chain state ));
+
+         assert_json_nonempty_list( $response->{auth_chain} );
+         my @auth_chain = @{ $response->{auth_chain} };
+
+         log_if_fail "Auth chain", \@auth_chain;
+
+         foreach my $event ( @auth_chain ) {
+            assert_is_valid_pdu( $event );
+            $event->{room_id} eq $room_id or
+               die "Expected auth_event room_id to be $room_id";
+         }
+
+         # Annoyingly, the "auth chain" isn't specified to arrive in any
+         # particular order. We'll have to keep walking it incrementally.
+
+         my %accepted_authevents;
+         while( @auth_chain ) {
+            my @accepted = extract_by {
+               $inbound_server->auth_check_event( $_, \%accepted_authevents )
+            } @auth_chain;
+
+            unless( @accepted ) {
+               log_if_fail "Unacceptable auth chain", \@auth_chain;
+
+               die "Unable to find any more acceptable auth_chain events";
+            }
+
+            $accepted_authevents{$_->{event_id}} = $_ for @accepted;
+         }
+
+         assert_json_nonempty_list( $response->{state} );
+         my %state = partition_by { $_->{type} } @{ $response->{state} };
+
+         log_if_fail "State", \%state;
+
+         # TODO: lots more checking. Requires spec though
+         Future->done(1);
+      });
+   };
+
+
+
+test "Outbound federation falls back to v1 /send_join if v2 isn't available",
```

Comment:
This is presumably the same as the outbound v2 /send_join test, but only awaiting on v1? Can we dedupe as per above?